README.md
<img width="300" src="https://files-ncvscphfon.now.sh/" alt="Andive UI Library" />

Style `<div />` the Ambler way.

[Storybook](https://andive.ambler.fr)

# Getting started

- `yarn add @ambler/andive react react-dom styled-components`

## Any React project

1. Render the <Baseline /> component which reset the SPA Css.

```
import {Baseline} from '@ambler/andive'
import App from './app'

function HelloWorld() {
  return (
    <>
      <Baseline />
      <App />
    </>
  )
}
```

2. Import the Asap Google font (default Andive's font).

```
<head>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Asap" />
</head>
```

3. Enjoy the set of components and rules andive provides :)

# API

To style our components we use `styled-components`. To avoid dependency collisions and let you configure your `styled-components` it is set as a _peerDependency_

## Typography

TODO

## Components

TODO

# Contribution

## How to release

If you are working on andive and want to share some unstable future features or refactor, do it from the canary channel. The canary channel mirrors the stable one.

### Canary workflow

Manually you:

- checkout the canary branch
- commit some changes
- `yarn release:canary`
- `git push origin canary --follow-tags`

Then CircleCI does:

- build the package
- TODO: test (removed because no test exists as of now)
- push to npm with the @canary tag

### Production workflow

Manullay you:

- checkout the master branch
- merge canary, when ready and stable, into master
- `yarn release`
- `git push origin master --follow-tags`

Then CircleCI does:

- build the package
- push to npm with the default @latest tag

### Publish the storybook

After a release is done, deploy the according storybook with:

- `yarn share`

- if canary: `now alias <outputed-now-build-url> andive-canary.ambler.fr`
- else if prod: `now alias <outputed-now-build-url> andive.ambler.fr`

## Project dependencies and setup

- Prettier
  VSCode: we let eslint format the code through eslint. Keep `editor.formatOnSave` disabled to let
  `eslint.autoFixOnSave` do his job.
- Eslint
  `eslint` uses prettier to format and disable its rules. This can be done through `eslint-plugin-prettier` and `eslint-config-prettier`,
  yet `eslint-plugin-prettier` exposes a "recommended" config that turns on both dependencies. See https://prettier.io/docs/en/eslint.html#use-both for more details.

  `babel-eslint` is the parser used to lint.

  `eslint-plugin-react` needs a configuration under `settings.react` key in `.eslintrc` to a least specify the react version. See https://github.com/yannickcr/eslint-plugin-react#configuration.

- Babel
  babel is used to transpile our components to commonjs format.

- Package versions are fixed with exact-version (except for peerDependencies because we want react and react-dom to match any minor update)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.11.0-canary.23](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.22...v0.11.0-canary.23) (2019-10-22)


### Bug Fixes

* **accordeon:** fix without icon padding ([0ceff00](https://github.com/amblerhq/ambler/commit/0ceff00))



## [0.11.0-canary.22](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.21...v0.11.0-canary.22) (2019-10-22)


### Bug Fixes

* **accordeon:** fix padding, icon alignment, and controlled mode ([5c24f35](https://github.com/amblerhq/ambler/commit/5c24f35))



## [0.11.0-canary.21](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.20...v0.11.0-canary.21) (2019-10-21)


### Bug Fixes

* **button:** pass button-group props through context ([6e7e69b](https://github.com/amblerhq/ambler/commit/6e7e69b))
* **checkbox:** update cursor on disabled ([c3d8ca4](https://github.com/amblerhq/ambler/commit/c3d8ca4))



## [0.11.0-canary.20](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.19...v0.11.0-canary.20) (2019-10-21)


### Bug Fixes

* **checkbox:** fix disabled state ([b35e3f7](https://github.com/amblerhq/ambler/commit/b35e3f7))



## [0.11.0-canary.19](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.18...v0.11.0-canary.19) (2019-10-13)


### Bug Fixes

* **select:** fix select native width overflow ([04a399a](https://github.com/amblerhq/ambler/commit/04a399a))



## [0.11.0-canary.18](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.17...v0.11.0-canary.18) (2019-10-13)


### Bug Fixes

* **dropdown:** show value even if disabled ([c19497f](https://github.com/amblerhq/ambler/commit/c19497f))
* **dropdown-menu:** fix buttonComponent prop-type ([6e2997a](https://github.com/amblerhq/ambler/commit/6e2997a))
* **dropdown-menu:** fix mobile fullWidth + add openVariant ([425f20d](https://github.com/amblerhq/ambler/commit/425f20d))



## [0.11.0-canary.17](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.16...v0.11.0-canary.17) (2019-10-09)


### Features

* **dropdown-filter:** add openLeft prop ([b2558a5](https://github.com/amblerhq/ambler/commit/b2558a5))



## [0.11.0-canary.16](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.15...v0.11.0-canary.16) (2019-10-09)


### Features

* **dropdown-menu:** add openLeft prop ([8fd992d](https://github.com/amblerhq/ambler/commit/8fd992d))



## [0.11.0-canary.15](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.14...v0.11.0-canary.15) (2019-10-08)


### Features

* **accordeon:** add Hover overflow prop ([30dd185](https://github.com/amblerhq/ambler/commit/30dd185))



## [0.11.0-canary.14](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.13...v0.11.0-canary.14) (2019-10-08)


### Features

* **accordeon:** add link variant ([cd05e92](https://github.com/amblerhq/ambler/commit/cd05e92))



## [0.11.0-canary.13](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.12...v0.11.0-canary.13) (2019-10-03)


### Features

* **info:** add disabled variant to info block ([cdaafd5](https://github.com/amblerhq/ambler/commit/cdaafd5))



## [0.11.0-canary.12](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.10...v0.11.0-canary.12) (2019-10-02)


### Bug Fixes

* **info:** Info.Block can have more then 1 child ([08b2abb](https://github.com/amblerhq/ambler/commit/08b2abb))


### Features

* **dropdown-menu:** add disabled prop to options ([3ec3a6f](https://github.com/amblerhq/ambler/commit/3ec3a6f))



## [0.11.0-canary.11](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.10...v0.11.0-canary.11) (2019-09-26)



## [0.11.0-canary.10](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.9...v0.11.0-canary.10) (2019-09-09)


### Bug Fixes

* **input:** fix placeholder text color ([e5b75a8](https://github.com/amblerhq/ambler/commit/e5b75a8))



## [0.11.0-canary.9](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.8...v0.11.0-canary.9) (2019-09-09)


### Bug Fixes

* **icons:** swap seated and lying icons ([a4f1eb4](https://github.com/amblerhq/ambler/commit/a4f1eb4))



## [0.11.0-canary.8](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.7...v0.11.0-canary.8) (2019-08-16)


### Bug Fixes

* **icon:** fix typo in question icon import ([e3b4b49](https://github.com/amblerhq/ambler/commit/e3b4b49))



## [0.11.0-canary.7](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.6...v0.11.0-canary.7) (2019-08-16)


### Bug Fixes

* **button:** rework buttons and their stories ([985f09e](https://github.com/amblerhq/ambler/commit/985f09e))
* **color:** add mainText + secondaryText colors ([8521ceb](https://github.com/amblerhq/ambler/commit/8521ceb))
* **theme-provider:** allow for multiple children ([7035781](https://github.com/amblerhq/ambler/commit/7035781))



## [0.11.0-canary.6](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.5...v0.11.0-canary.6) (2019-08-14)


### Bug Fixes

* **dropdown:** fix disabled color ([65da2c5](https://github.com/amblerhq/ambler/commit/65da2c5))


### Tests

* add visual regression testing ([e013023](https://github.com/amblerhq/ambler/commit/e013023))



## [0.11.0-canary.5](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.4...v0.11.0-canary.5) (2019-08-14)


### Features

* **space:** add vspace and hspace components ([9533226](https://github.com/amblerhq/ambler/commit/9533226))



## [0.11.0-canary.4](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.3...v0.11.0-canary.4) (2019-08-14)


### Bug Fixes

* **colors:** fix colors and icon colors ([60e41a7](https://github.com/amblerhq/ambler/commit/60e41a7))



## [0.11.0-canary.3](https://github.com/amblerhq/ambler/compare/v0.11.0-canary.2...v0.11.0-canary.3) (2019-08-14)


### Features

* **icon:** add incident icon ([5a89a75](https://github.com/amblerhq/ambler/commit/5a89a75))

src/components/button.js
import React from 'react'
import styled, {css} from 'styled-components'
import PropTypes from 'prop-types'

import * as palette from '../constants/palette'
import {Body2, body1Css, body3Css} from './typography'
import Loader from './loader'

const lightGreyAlpha = palette.hexToRGBA(palette.mediumPrimary, 0.5)
const lightBeetrootPurpleAlpha = palette.hexToRGBA(palette.mediumBeetrootPurple, 0.5)
const mediumBeetrootPurpleAlpha = palette.hexToRGBA(palette.darkBeetrootPurple, 0.5)

const ButtonRoot = styled.div.attrs({
  'data-andive-type': 'button'
})`
  padding: ${props => props.theme.padding}px;
  cursor: ${props => (props.disabled ? 'not-allowed' : props.loading ? 'progress' : 'pointer')};
`

const ResetButton = styled.button`
  outline: none;
  border: none;
  background: none;
  cursor: inherit;

  display: flex;
  justify-content: center;
  align-items: center;

  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`

const DefaultButtonText = styled(Body2)`
  text-transform: uppercase;
  font-size: ${props => (props.mobile ? 14 : 16)}px;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
`

// * Default/secondary variant

function defaultColor({disabled}) {
  if (disabled) {
    return lightBeetrootPurpleAlpha
  }

  return palette.mediumBeetrootPurple
}

const DefaultButton = styled(ResetButton)`
  border: 1px solid ${palette.mediumBeetrootPurple};
  background-color: ${props => props.backgroundColor || palette.white};
  color: ${props => props.color};
  height: ${props => (props.mobile ? 40 : 56)}px;
  border-radius: 28px;
  box-shadow: 0 2px 5px 0 rgba(5, 71, 82, 0.2);

  padding: ${props => (props.mobile ? '8px 16px' : '16px 24px')};

  min-width: 200px;
  ${props =>
    (props.minWidth || props.small) &&
    css`
      min-width: ${props => (props.small ? 0 : props.minWidth)}px;
    `}

  ${props =>
    props.disabled &&
    css`
      background-color: ${palette.lightGrey};
      border: 1px solid ${lightGreyAlpha};
      box-shadow: none;
    `}

  ${props =>
    props.loading &&
    css`
      padding: 0;
      min-width: ${props => (props.mobile ? 40 : 56)}px;
    `}
`

// * Primary/main variant

function primaryColor({disabled}) {
  if (disabled) {
    return palette.hexToRGBA(palette.white, 0.6)
  }

  return palette.white
}

const PrimaryButton = styled(DefaultButton)`
  border: 1px solid ${props => (props.invert ? palette.white : palette.mediumBeetrootPurple)};
  background-color: ${props => props.backgroundColor || palette.mediumBeetrootPurple};
  color: ${props => props.color};
  border-radius: 28px;
  box-shadow: 0 2px 5px 0 ${palette.hexToRGBA(palette.darkPrimary, 0.2)};

  ${props =>
    props.disabled &&
    css`
      background: ${lightBeetrootPurpleAlpha};
      border: 1px solid ${({invert}) => (invert ? lightGreyAlpha : 'transparent')};
      box-shadow: none;
    `}
`

// * Flat variant

function flatColor({disabled, invert}) {
  if (disabled) {
    return invert ? palette.hexToRGBA(palette.white, 0.6) : palette.hexToRGBA(palette.mediumBerryBlue, 0.6)
  }

  return invert ? palette.white : palette.mediumBerryBlue
}

const FlatButton = styled(ResetButton)`
  color: ${props => props.color};
`

// * Filter variant

function filterColor({disabled, invert}) {
  if (disabled) {
    return invert ? lightGreyAlpha : mediumBeetrootPurpleAlpha
  }

  return invert ? palette.darkPrimary : palette.white
}

const FilterButton = styled(ResetButton)`
  /* The padding is ruled by the text, not the button. */
  padding: 0;

  border-radius: 20px;
  border: 1px solid ${props => (props.invert ? palette.darkPrimary : 'transparent')};
  height: ${props => (props.mobile ? 32 : 40)}px;
  background: ${({invert, backgroundColor}) => backgroundColor || (invert ? palette.white : palette.mediumBerryBlue)};
`

const FilterText = styled.div`
  ${props => (props.mobile ? body3Css : body1Css)};

  color: ${props => props.color};
  padding: ${props => (props.mobile ? '6px 18px' : '8px 24px')};
`

function textStyle(leftIcon, rightIcon) {
  return {
    paddingRight: rightIcon ? 8 : undefined,
    paddingLeft: leftIcon ? 8 : undefined
  }
}

const Button = React.forwardRef(function Button(
  {
    label,
    rightIcon,
    leftIcon,
    onClick,
    variant,
    invert,
    textColor,
    backgroundColor,
    disabled,
    loading,
    mobile,
    ...props
  },
  ref
) {
  const ButtonComponent = React.useMemo(() => {
    switch (variant) {
      case 'primary':
        return PrimaryButton
      case 'flat':
        return FlatButton
      case 'filter':
        return FilterButton
      default:
        return DefaultButton
    }
  }, [variant])

  const ButtonLabel = React.useMemo(() => {
    switch (variant) {
      case 'flat':
        return Body2
      case 'filter':
        return FilterText
      case 'primary':
      default:
        return DefaultButtonText
    }
  }, [variant])

  const colorFn = React.useMemo(
    () =>
      variant === 'primary'
        ? primaryColor
        : variant === 'flat'
        ? flatColor
        : variant === 'filter'
        ? filterColor
        : defaultColor,
    [variant]
  )

  const color = React.useMemo(() => {
    const color = colorFn({disabled, invert})
    if (disabled) {
      return color
    }

    // ? The textColor override takes precedence over the normal color, not the disabled one.
    if (textColor) {
      return textColor
    }

    return color
  }, [variant, disabled, invert, textColor])

  const loaderColor = React.useMemo(() => {
    if (!loading) {
      return
    }

    switch (variant) {
      case 'primary':
        return palette.white
      case 'flat':
        return invert ? palette.white : palette.mediumBerryBlue
      default:
        return palette.mediumBeetrootPurple
    }
  }, [variant, loading, invert])

  return (
    <ButtonRoot disabled={disabled} loading={loading}>
      <ButtonComponent
        ref={ref}
        onClick={!disabled && !loading ? onClick : undefined}
        type="button"
        primary={variant === 'primary'}
        invert={invert}
        disabled={disabled}
        loading={loading}
        mobile={mobile}
        backgroundColor={backgroundColor}
        color={color}
        {...props}
      >
        {loading ? (
          <Loader inline color={loaderColor} />
        ) : (
          <>
            {leftIcon && React.cloneElement(leftIcon, {color})}
            <ButtonLabel style={textStyle(leftIcon, rightIcon)} invert={invert} color={color} mobile={mobile}>
              {label}
            </ButtonLabel>
            {rightIcon && React.cloneElement(rightIcon, {color})}
          </>
        )}
      </ButtonComponent>
    </ButtonRoot>
  )
})

Button.propTypes = {
  label: PropTypes.string,
  rightIcon: PropTypes.node,
  leftIcon: PropTypes.node,
  onClick: PropTypes.func,
  /** Change the button look to "primary", "flat", or undefined (defautlt) */
  variant: PropTypes.oneOf(['primary', 'flat', 'filter']),
  /** Inverse button colors to display on beetrootPurple background */
  invert: PropTypes.bool,
  backgroundColor: PropTypes.string,
  textColor: PropTypes.string,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  mobile: PropTypes.bool,
  /** See ButtonGroup for this prop usage. Do not pass this props directly ! */
  minWidth: PropTypes.number
}

export default Button
